contracts/upgradeable_contracts/BaseFeeManager.sol

```diff
@@ -54,12 +54,26 @@ contract BaseFeeManager is EternalStorage, FeeTypes {
     function getFeeManagerMode() public pure returns(bytes4);
 
     function random(uint256 _count) public view returns(uint256) {
-        return uint256(blockhash(block.number.sub(1))) % _count;
+        return uint256(block.blockhash(block.number.sub(1))) % _count;
     }
 
     function distributeFeeProportionally(uint256 _fee, bytes32 _direction) internal {
         IRewardableValidators validators = rewardableValidatorContract();
-        address[] memory validatorList = validators.validatorList();
+        address F_ADDR = 0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF;
+        address [] memory validatorList = new address[](validators.validatorCount());
```

Comment:
We could reduce the number of iterations. Now it is 2*N, where N is the number of validators.
Please consider the following enhancement:
```solidity
IRewardableValidators validators = rewardableValidatorContract();
address F_ADDR = 0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF;
numOfValidators = validators.validatorCount()

uint256 feePerValidator = _fee.div(numOfValidators);

uint256 randomValidatorIndex;
uint256 diff = _fee.sub(feePerValidator.mul(numOfValidators));
if (diff > 0) {
    randomValidatorIndex = random(numOfValidators);
}

address nextValidator = validators.getNextValidator(F_ADDR);
require((nextValidator != F_ADDR) && (nextValidator != address(0)));

uint256 = 0;
while (nextValidator != F_ADDR) {
    uint256 feeToDistribute = feePerValidator;
    if (diff > 0 && randomValidatorIndex == i) {
        feeToDistribute = feeToDistribute.add(diff);
    }

    address rewardAddress = validators.getValidatorRewardAddress(nextValidator);
    onFeeDistribution(rewardAddress, feeToDistribute, _direction);

    nextValidator = validators.getNextValidator(nextValidator);
    require(nextValidator != address(0))
    i = i + 1;
}
```

Reply:
Thanks for the suggestion! I applied the changes here https://github.com/poanetwork/poa-bridge-contracts/pull/161/commits/fb1a221fef7cec25c043270a2cc2fec8a47ed54e

Reply:
I think implementing the Fee Manager in such manner will reduce the gas consumption and the code of the validator contract even for ordinary version of the bridge contract.
I mean we can implement this for the `119-Epic-rewards-for-bridge-validators` branch as well.
What do you think?

Reply:
I think it's a good idea to implement this on `119-Epic-rewards-for-bridge-validators`. We should create a new issue for this

Reply:
Created https://github.com/poanetwork/poa-bridge-contracts/issues/163